### DISCUSSION TITLE
How to assign to a pydantic AnyUrl?

### DISCUSSION THREAD

**Discussion by mjperrone** [microsoft/pyright#8664] [Q&A]
I'm trying to instantiate a pydantic model which contains a field that is an `AnyUrl`. The few ways I've tried both yield pyright errors `reportArgumentType`. Is there a correct way to do this?

Here is a minimal repro. This uses pydantic version `1.10.15` and pyright version `1.1.374`.

```python
class Example(BaseModel):
    anyUrl: AnyUrl = Field(
        ..., description="description"
    )
ex = Example(anyUrl="https://github.com/fakeorg/fakerepo.git")
ex = Example(anyUrl=AnyUrl("https://github.com/fakeorg/fakerepo.git"))
```

```
error: Argument of type "Literal['https://github.com/fakeorg/fakerepo.git']" cannot be assigned to parameter "anyUrl" of type "AnyUrl" in function "__init__"
    "Literal['https://github.com/fakeorg/fakerepo.git']" is incompatible with "AnyUrl" (reportArgumentType)
error: Argument of type "object" cannot be assigned to parameter "anyUrl" of type "AnyUrl" in function "__init__"
    "object" is incompatible with "AnyUrl" (reportArgumentType)
```

**Comment by mjperrone** [comment]
@erictraut is this something you can advise on? Or do you think I should take this up with the pydantic folks?

**Reply by erictraut** [reply]
I don't know enough about pydantic or `AnyUrl` to offer any advice here. I can say that pyright is working correctly here based on the type information it is given.

**Comment by Viicos** [comment]
@mjperrone this is a limitation of the [`dataclass_transform`](https://typing.readthedocs.io/en/latest/spec/dataclasses.html#the-dataclass-transform-decorator) decorator. `AnyUrl` is defined as a Protocol class in Pydantic, but the library will perform the necessary type casting if you provide a string. The same will happen with a simplified example:

```python
from pydantic import BaseModel

class Model(BaseModel):
    foo: int

Model(foo='1')  # type error, fine at runtime
#> Model(foo=1)
```

You can ignore the error or alternatively use the `model_validate` method: `Model.model_validate({'foo': '1'})` (but you don't get type checking of the argument names).
